examples/move/scratch_off/sources/game.move

```diff
@@ -0,0 +1,256 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+
+/// This is an example implementation of a scratch off game utilizing a BLS signature to generate randomness.
+/// The game consists of a store which is a shared_object where users can go to evaluate their tickets.
+/// A ticket object that is transferable and can be passed into be evaluated.
+module scratch_off::game {
+
+    use std::vector;
+    use sui::coin::{Self, Coin};
+    use sui::balance::{Self, Balance};
+    use sui::object::{Self, UID, ID};
+    use sui::tx_context::{Self, TxContext};
+    use sui::dynamic_object_field as dof;
+    use sui::transfer;
+    use sui::event;
+    use sui::bls12381::bls12381_min_pk_verify;
+    use sui::hash::blake2b256;
+    use scratch_off::math;
+
+    // --------------- Constants ---------------
+    const DEFAULT_TOTAL_ODDS: u64 = 10000;
+
+    const EInvalidInputs: u64 = 0;
+    const EInvalidBlsSig: u64 = 1;
+    const ENoTicketsLeft: u64 = 2;
+
+    // --------------- Events ---------------
+    struct NewDrawing<phantom T> has copy, drop {
+        ticket_id: ID,
+        player: address,
+    }
+
+    struct DrawingResult<phantom T> has copy, drop {
+        ticket_id: ID,
+        player: address,
+        amount_won: u64,
+    }
+
+    struct Ticket has key, store {
+        id: UID,
+        /// Which store this ticket belongs to
+        convenience_store_id: ID,
+        /// Original receiver address, but on evaluation this is set to the ctx::sender addr
+        player: address,
+	}
+
+    struct PrizeStruct has store {
+        ticket_amount: u64,
+        prize_value: u64,
+    }
+    
+	struct ConvenienceStore<phantom Asset> has key {
+        id: UID,
+        creator: address,
+        /// Total prize pool available.
+        prize_pool: Balance<Asset>,
+        /// The vector of tickets and their corresponding prize winnings in the pool
+        /// of winning tickets. (tickets left, prize awarded)
+        winning_tickets: vector<PrizeStruct>,
+        /// Total number of losing tickets
+        losing_tickets_left: u64,
+        /// Total number of winning tickets
+        winning_tickets_left: u64,
+        /// Total number of tickets originally available
+        original_ticket_count: u64,
+        /// Total number of tickets issued
+        tickets_issued: u64,
+        public_key: vector<u8>,
+	}     
+
+    public fun total_tickets<Asset>(store: &ConvenienceStore<Asset>): u64 {
+        store.winning_tickets_left + store.losing_tickets_left
+    }
+
+    /// Initializes the store with all of the lottery tickets.
+    /// We allow the user of the store.
+    /// We purposely design the convenience store to be an owned object so that we
+    /// don't need to make this in a shared format.
+    public fun open_store<Asset>(
```

Comment:
should we limit calls to send_ticket and evaluate_ticket by time?

Reply:
send ticket will also be called only by us. We should have a limit at some point, but I was thinking we can also just stop the game when the balance is 0

Reply:
sure, I don't have strong preference here, just wanted to flag that as an option (as the "game theory" is different for those options - if I know that time is short, I might play differently).